### PR TITLE
Optimize memory allocations (#98702)

### PR DESCRIPTION
**Pull request by cd21h**
Use `ArrayPool` to reduce memory allocations when writing JSON

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs [hunk 1]
```diff
@@ -76,7 +79,17 @@ public override void Write<TState>(in LogEntry<TState> logEntry, IExternalScopeP
                     writer.Flush();
                 }
 #if NETCOREAPP
-                textWriter.Write(Encoding.UTF8.GetString(output.WrittenMemory.Span));
+                var messageBytes = output.WrittenMemory.Span;
+                var logMessageBuffer = ArrayPool<char>.Shared.Rent(Encoding.UTF8.GetCharCount(messageBytes));
```
Comment (MichalPetryka) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs:83]: You probably want `GetMaxCharCount` here as `GetCharCount` is relatively expensive.

Comment (shatl) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs:83]: `GetMaxCharCount` is a worst case, which will result in (over) allocating large buffers for long string.

Comment (MichalPetryka) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs:83]: `GetCharCount` on the other hand is quite slow and takes almost as much as doing `GetChars` IIRC. Since we're pooling, overallocating probably shouldn't matter here.

Comment (shatl) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs]: @MichalPetryka  Updated

**Comment by shatl** [comment]
@MichalPetryka  Could you take a look, please?

File: src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs [hunk 2]
```diff
@@ -76,7 +79,17 @@ public override void Write<TState>(in LogEntry<TState> logEntry, IExternalScopeP
...
+                {
+                    var charsWritten = Encoding.UTF8.GetChars(messageBytes, logMessageBuffer);
+                    textWriter.Write(logMessageBuffer, 0, charsWritten);
+                }
+                finally
+                {
+                    ArrayPool<char>.Shared.Return(logMessageBuffer);
+                }
 #else
                 textWriter.Write(Encoding.UTF8.GetString(output.WrittenMemory.Span.ToArray()));
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs:94]: >output.WrittenMemory.Span.ToArray() [](http://example.com/codeflow?start=57&length=35)

This part can be optimized too by using the unsafe version of the [GetString](https://learn.microsoft.com/en-us/dotnet/api/system.text.encoding.getstring?view=net-8.0#system-text-encoding-getstring(system-byte*-system-int32)). It will save at least one allocation.

**Comment by tarekgh** [comment]
@cd21h could you please resolve the source conflict to be able to proceed? 

Also, for my comment https://github.com/dotnet/runtime/pull/99542#pullrequestreview-2053489708 feel free to address it or if you want can address it in another PR. Thanks!

**Comment by tarekgh** [comment]
@cd21h I have pushed a commit for optimizing the other case in addition to merging main to your branch. Please have a look and let me know if it makes sense to you. Also, could you please accept the license agreement too https://github.com/dotnet/runtime/pull/99542#issuecomment-2109061793? Thanks!

**Comment by cd21h** [comment]
> @cd21h please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @dotnet-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @dotnet-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @dotnet-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@dotnet-policy-service agree

**Comment by cd21h** [comment]
@tarekgh Looks good to me!

File: src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs [hunk 3]
```diff
@@ -81,11 +84,30 @@ public override void Write<TState>(in LogEntry<TState> logEntry, IExternalScopeP
...
+                    var charsWritten = Encoding.UTF8.GetChars(messageBytes, logMessageBuffer);
+ #else
+                    int charsWritten;
+                    unsafe
+                    {
+                        fixed (byte* messageBytesPtr = messageBytes)
+                        fixed (char* logMessageBufferPtr = logMessageBuffer)
+                        {
+                            charsWritten = Encoding.UTF8.GetChars(messageBytesPtr, messageBytes.Length, logMessageBufferPtr, logMessageBuffer.Length);
+            }
```
Comment (MihaZupan) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs:102]:
```suggestion
                        }
```

Comment (cd21h) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs]: done

File: src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs [hunk 4]
```diff
@@ -2,6 +2,9 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+#if NETCOREAPP
```
Comment (tarekgh) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs:5]: Why is this needed?

Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs:5]: This is causing the failure `src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs(89,40): error CS0103: The name 'ArrayPool' does not exist in the current context`.

Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs]: I pushed the fix.

**Review by tarekgh — APPROVED** [review]
